### PR TITLE
Feature/#30 build the data science sandbox as docker image

### LINKED ISSUE #30: Build the Data Science Sandbox as Docker Image

### PR DESCRIPTION
**Pull request by ckunki**
closes #30

### CODE REVIEW

File: test/integration/test_create_dss_docker_image.py [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+    http_conn = requests.get(f"http://{ip_address}:8888/lab")
+    assert http_conn.status_code == 200
+
+
+def test_install_notebook_connector(dss_docker_container):
+    container = dss_docker_container
+    command = '/root/jupyterenv/bin/python -c "import exasol.secret_store"'
+    exit_code, output = container.exec_run(command)
+    output = output.decode('utf-8').strip()
+    assert exit_code == 0, f'Got output "{output}".'
```
Comment (kaklakariada) [inline test/integration/test_create_dss_docker_image.py:68]: Does it make sense to verify a certain output content?

File: test/unit/test_pretty_print.py [hunk 2]
```diff
@@ -0,0 +1,30 @@
+import re
+import time
+from exasol.ds.sandbox.lib import pretty_print
+from datetime import datetime, timedelta
+
+
+def test_elapsed_rounded():
+    start = datetime.now()
+    time.sleep(1.2)
```
Comment (kaklakariada) [inline test/unit/test_pretty_print.py:9]: Sleeping in a simple unit test feels a waste of time. Could you change the API to avoid this?

**Review by kaklakariada — CHANGES_REQUESTED** [review]

File: exasol/ds/sandbox/cli/commands/create_docker_image.py [hunk 3]
```diff
@@ -0,0 +1,32 @@
+import click
+
+from exasol.ds.sandbox.cli.cli import cli
+from exasol.ds.sandbox.cli.common import add_options
+from exasol.ds.sandbox.lib.dss_docker import DssDockerImage
+
+
+_options = [
```
Comment (tkilias) [inline exasol/ds/sandbox/cli/commands/create_docker_image.py:8]: Why don't you add the options not directly to the function?

File: exasol/ds/sandbox/cli/commands/create_docker_image.py [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+        help="Organization and repository on hub.docker.com to publish the docker image to."),
+    click.option(
+        '--publish', type=bool, default=False, show_default=True,
+        help="Whether to publish the created Docker image"),
+]
+
+
+@cli.command()
+@add_options(_options)
+def create_docker_image(repository: str, publish: bool):
```
Comment (tkilias) [inline exasol/ds/sandbox/cli/commands/create_docker_image.py:21]: How do you define the tag? The default tag should be the project version, but for testing it is probably a good idea to be able to specify the tag.

File: exasol/ds/sandbox/cli/commands/create_docker_image.py [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+    """
+    Create a Docker image for data-science-sandbox and deploy
+    it to https://hub.docker.com/exasol/data-science-sandbox.
+    """
+    print("Hello this is create_docker_image, using")
```
Comment (tkilias) [inline exasol/ds/sandbox/cli/commands/create_docker_image.py:26]: Debug output? Otherwise, please have a look at the other commands, how they do logging

File: exasol/ds/sandbox/lib/dss_docker/create_image.py [hunk 6]
```diff
@@ -0,0 +1,99 @@
+import docker
+import logging
+from datetime import datetime
+from docker.types import Mount
+from exasol.ds.sandbox.lib import pretty_print
+from importlib_metadata import version
+from pathlib import Path
+
+from exasol.ds.sandbox.lib.config import ConfigObject, SLC_VERSION
+# renaming proposal:
```
Comment (tkilias) [inline exasol/ds/sandbox/lib/dss_docker/create_image.py:10]: Please create a ticket, that will be a shotgun change and should go into its own PR. And, remove the comment afterwards.

File: exasol/ds/sandbox/lib/dss_docker/create_image.py [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+        }
+        return AnsibleRunContext(
+            playbook="dss_docker_playbook.yml",
+            extra_vars=extra_vars,
+        )
+
+    def _ansible_config(self) -> ConfigObject:
+        return ConfigObject(
+            time_to_wait_for_polling=0.1,
+            slc_version=SLC_VERSION,
```
Comment (tkilias) [inline exasol/ds/sandbox/lib/dss_docker/create_image.py:50]: Do we still need the SLC_VERSION? Or, is it only partially removed, yet?

File: exasol/ds/sandbox/lib/dss_docker/create_image.py [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+        )
+
+    def create(self):
+        logger = logging.getLogger(__name__)
+        logger.setLevel(self.log_level)
+        logging.basicConfig(format="%(asctime)s %(levelname)s %(message)s")
```
[the remark this answers is not in the document]
Comment (tkilias) [reply exasol/ds/sandbox/lib/dss_docker/create_image.py:56]: Logging config should happen in cli and not in lib

File: exasol/ds/sandbox/lib/dss_docker/create_image.py [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+        try:
+            start = datetime.now()
+            docker_client = docker.from_env()
+            path = Path(__file__).parent
```
Comment (tkilias) [inline exasol/ds/sandbox/lib/dss_docker/create_image.py:60]: Use importlib resources to get the docker file

File: exasol/ds/sandbox/lib/dss_docker/create_image.py [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+            logger.info(
+                f"Creating docker image {self.image_name}"
+                f" from {path / 'Dockerfile'}"
+            )
+            docker_client.images.build(path=str(path), tag=self.image_name)
+            socket_mount = Mount("/var/run/docker.sock", "/var/run/docker.sock", type="bind")
```
Comment (tkilias) [inline exasol/ds/sandbox/lib/dss_docker/create_image.py:66]: Mount shouldn't be needed during build

File: exasol/ds/sandbox/lib/dss_docker/create_image.py [hunk 11]
```diff
@@ -0,0 +1,99 @@
...
+            mapped_ports = {'8888/tcp': 8888}
```
Comment (tkilias) [inline exasol/ds/sandbox/lib/dss_docker/create_image.py:67]: Mapped ports shouldn't be needed during build

Comment (tkilias) [reply exasol/ds/sandbox/lib/dss_docker/create_image.py:67]: However, you might want to add this as exposed ports to the docker file

https://docs.docker.com/engine/reference/builder/#expose

File: exasol/ds/sandbox/lib/dss_docker/create_image.py (hunk 10 above)
Comment (tkilias) [reply exasol/ds/sandbox/lib/dss_docker/create_image.py:66]: However, you might want to add it to the dockerfile

https://docs.docker.com/engine/reference/builder/#volume

File: exasol/ds/sandbox/lib/dss_docker/create_image.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (ckunki) [reply exasol/ds/sandbox/lib/dss_docker/create_image.py:56]: Discussed, left for potential future refactoring

File: exasol/ds/sandbox/lib/dss_docker/create_image.py [hunk 12]
```diff
@@ -0,0 +1,99 @@
...
+            image = container.commit(
+                repository=self.image_name,
+            )
+        except Exception as ex:
+            raise ex
+        finally:
+            logger.info("Stopping container")
+            container.stop()
+            logger.info("Removing container")
+            container.remove()
```
Comment (tkilias) [inline exasol/ds/sandbox/lib/dss_docker/create_image.py:96]: Container removal should be configurable for debugging

File: test/integration/test_create_dss_docker_image.py (hunk 1 above)
Comment (ckunki) [reply test/integration/test_create_dss_docker_image.py:68]: Command currently has not output, Variable `output` is only used in case of test failure

File: exasol/ds/sandbox/lib/pretty_print.py [hunk 13]
```diff
@@ -0,0 +1,23 @@
+from datetime import datetime, timedelta
+
+
+def elapsed(start: datetime, round_to_seconds=True) -> str:
```
Comment (tkilias) [inline exasol/ds/sandbox/lib/pretty_print.py:4]: Their is package Stopwatch which might help with that. We use it already in other packages.

File: exasol/ds/sandbox/lib/pretty_print.py [hunk 14]
```diff
@@ -0,0 +1,23 @@
...
+    d = datetime.now() - start
+    if round_to_seconds:
+        d = d - timedelta(microseconds=d.microseconds)
+    return str(d)
+
+
+def size(num: int, factor: float = 1024, unit: str = "B") -> str:
```
Comment (tkilias) [inline exasol/ds/sandbox/lib/pretty_print.py:11]: There is package called humanfriendly which might help here. We use it already in other packages

File: exasol/ds/sandbox/runtime/ansible/dss_docker_playbook.yml [hunk 15]
```diff
@@ -0,0 +1,22 @@
...
+  vars:
+      ansible_python_interpreter: python3
+  tasks:
+    - name: Add docker container to inventory
+      add_host:
+        name: "{{docker_container}}"
+        ansible_connection: docker
+
+- name: Setup DSS Docker Container
+  hosts: ds-sandbox-docker # cannot use variable here
```
Comment (tkilias) [inline exasol/ds/sandbox/runtime/ansible/dss_docker_playbook.yml:13]: Ok, this gets an interesting name, because if we need to fix this name, there are no concurrent runs allowed. Let's discuss this later.

File: test/integration/test_create_dss_docker_image.py [hunk 16]
```diff
@@ -0,0 +1,68 @@
...
+    )
+    print(
+        "\n- Using"
+        f' Docker container {testee.container_name}'
+        f' with image {testee.image_name}'
+    )
+    testee.create()
+    client = docker.from_env()
+    mapped_ports = {'8888/tcp': 8888}
+    container = client.containers.create(
```
Comment (tkilias) [inline test/integration/test_create_dss_docker_image.py:28]: You don't map the docker socket. I think, we need another cli/API for starting a container correctly. However, this is a new ticket.

File: test/integration/test_create_dss_docker_image.py [hunk 17]
```diff
@@ -0,0 +1,68 @@
...
+        version=f"{timestamp}",
+        publish=False,
+        log_level=logging.INFO,
+    )
+    print(
+        "\n- Using"
+        f' Docker container {testee.container_name}'
+        f' with image {testee.image_name}'
+    )
+    testee.create()
```
Comment (tkilias) [inline test/integration/test_create_dss_docker_image.py:25]: I would suggest, to move the image creation into its own fixture. That makes it easier to understand.

File: test/integration/test_create_dss_docker_image.py [hunk 18]
```diff
@@ -0,0 +1,68 @@
...
+    """
+    jupyter_command = (
+        "/root/jupyterenv/bin/jupyter-lab"
+        " --notebook-dir=/root/notebooks"
+        " --no-browser"
+        " --allow-root"
+    )
+    container = dss_docker_container
+    container.exec_run(jupyter_command, detach=True)
+    time.sleep(5.0)
```
Comment (tkilias) [inline test/integration/test_create_dss_docker_image.py:56]: I would instead poll a certain time the url and fail with a timeout if doesn't come up.

File: test/integration/test_create_dss_docker_image.py (hunk 1 above)
Comment (tkilias) [reply test/integration/test_create_dss_docker_image.py:68]: I think, there is not output, because we only import something

File: exasol/ds/sandbox/cli/commands/create_docker_image.py (hunk 4 above)
Comment (tkilias) [reply exasol/ds/sandbox/cli/commands/create_docker_image.py:21]: I think also the log level is missing

File: test/unit/test_pretty_print.py (hunk 2 above)
Comment (tkilias) [reply test/unit/test_pretty_print.py]: We should not implement that by our self, then the tests are also gone. See, my previous comments regarding the packages stopwatch and humanfriendly

**Review by tkilias — CHANGES_REQUESTED** [review]

File: exasol/ds/sandbox/lib/dss_docker/create_image.py [hunk 19]
```diff
@@ -50,28 +51,26 @@ def _ansible_config(self) -> ConfigObject:
             slc_version=SLC_VERSION,
         )
 
+    def _docker_file(self) -> Path:
+        return (
+            importlib_resources
```
Comment (tkilias) [inline exasol/ds/sandbox/lib/dss_docker/create_image.py:57]: This is not always a path it is a traversable. Usually, it is the best to read the content of the traversable which could point into a archive file. The docker python SDK for build supports fileobj, it is probably best to get a fileobj from the traversable and use this

File: exasol/ds/sandbox/lib/dss_docker/create_image.py [hunk 20]
```diff
@@ -50,28 +51,26 @@ def _ansible_config(self) -> ConfigObject:
...
+            .files("exasol.ds.sandbox.lib.dss_docker")
+            .joinpath("Dockerfile")
+        )
+
     def create(self):
-        logger = logging.getLogger(__name__)
-        logger.setLevel(self.log_level)
-        logging.basicConfig(format="%(asctime)s %(levelname)s %(message)s")
+        logger = get_status_logger(LogType.DOCKER_IMAGE)
```
Comment (tkilias) [inline exasol/ds/sandbox/lib/dss_docker/create_image.py:63]: I would create the logger outside of the class

File: test/integration/test_create_dss_docker_image.py [hunk 21]
```diff
@@ -1,33 +1,42 @@
...
+    #     f' with image {testee.image_name}'
+    # )
     testee.create()
+    try:
+        yield testee
+    finally:
+        docker.from_env().images.remove(testee.image_name)
+
+
+@pytest.fixture(scope="session")
```
Comment (tkilias) [inline test/integration/test_create_dss_docker_image.py:27]: To make this a session fixture needs to be discussed, because you might actually want a new container,  in case your test messes up the container. Starting a new container also should be a big time investment.

File: test/integration/test_create_dss_docker_image.py [hunk 22]
```diff
@@ -38,7 +47,23 @@ def dss_docker_container():
     finally:
         container.stop()
         container.remove()
-        client.images.remove(testee.image_name)
+
+
+def request_with_retry(url: str, timeout: timedelta, verbose:bool = False) -> requests.Response:
```
Comment (tkilias) [inline test/integration/test_create_dss_docker_image.py:46]: https://tenacity.readthedocs.io/en/latest/ already implements that, and we use it already in other projects

File: exasol/ds/sandbox/lib/pretty_print.py (hunk 13 above)
Comment (tkilias) [reply exasol/ds/sandbox/lib/pretty_print.py:4]: You only resolved the comment without fixing or answering. Please check again.

**Review by tkilias — CHANGES_REQUESTED** [review]

**Review by kaklakariada — DISMISSED** [review]

**Review by tkilias — APPROVED** [review]
